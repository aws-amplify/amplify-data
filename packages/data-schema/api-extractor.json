{
  "extends": "../../api-extractor.base.json",
  "mainEntryPointFilePath": "<projectFolder>/dist/esm/src/index.d.ts"
}
