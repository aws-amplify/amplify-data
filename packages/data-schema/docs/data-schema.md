<!-- Do not edit this file. It is automatically generated by API Documenter. -->

[Home](./index.md) &gt; [@aws-amplify/data-schema](./data-schema.md)

## data-schema package

## Enumerations

<table><thead><tr><th>

Enumeration


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[ModelRelationshipTypes](./data-schema.modelrelationshiptypes.md)


</td><td>

Model relationship types


</td></tr>
</tbody></table>

## Interfaces

<table><thead><tr><th>

Interface


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[EnumType](./data-schema.enumtype.md)


</td><td>

Enum type definition content


</td></tr>
</tbody></table>

## Namespaces

<table><thead><tr><th>

Namespace


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[a](./data-schema.a.md)


</td><td>


</td></tr>
</tbody></table>

## Type Aliases

<table><thead><tr><th>

Type Alias


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[Authorization](./data-schema.authorization.md)


</td><td>

Container for authorization schema definition content.


</td></tr>
<tr><td>

[ClientSchema](./data-schema.clientschema.md)


</td><td>


</td></tr>
<tr><td>

[CombinedModelSchema](./data-schema.combinedmodelschema.md)


</td><td>

CombinedModel schema definition interface


</td></tr>
<tr><td>

[CustomOperation](./data-schema.customoperation.md)


</td><td>

Custom operation definition interface


</td></tr>
<tr><td>

[CustomType](./data-schema.customtype.md)


</td><td>

Custom type container


</td></tr>
<tr><td>

[Json](./data-schema.json.md)


</td><td>

A precise, recursive Json type blows the type calculation stack without installing explicit `Json extends T ? short-circuit : ...` type checks all over the place. We may take that on later. But, this is a good-enough approximation for now.


</td></tr>
<tr><td>

[ModelDefaultIdentifier](./data-schema.modeldefaultidentifier.md)


</td><td>

Model default identifier


</td></tr>
<tr><td>

[ModelField](./data-schema.modelfield.md)


</td><td>

Public API for the chainable builder methods exposed by Model Field. The type is narrowing e.g., after calling .array() it will be omitted from intellisense suggestions


</td></tr>
<tr><td>

[ModelRelationshipField](./data-schema.modelrelationshipfield.md)


</td><td>

Model relationship field definition interface


</td></tr>
<tr><td>

[ModelRelationshipTypeArgFactory](./data-schema.modelrelationshiptypeargfactory.md)


</td><td>

Model relationship type definition content


</td></tr>
<tr><td>

[ModelSchema](./data-schema.modelschema.md)


</td><td>

Model schema definition interface


</td></tr>
<tr><td>

[ModelType](./data-schema.modeltype.md)


</td><td>

Model type definition interface


</td></tr>
<tr><td>

[Nullable](./data-schema.nullable.md)


</td><td>

Field type arg mutators


</td></tr>
<tr><td>

[RDSModelSchema](./data-schema.rdsmodelschema.md)


</td><td>

RDSModel schema definition interface


</td></tr>
<tr><td>

[RefType](./data-schema.reftype.md)


</td><td>

Reference type definition interface


</td></tr>
<tr><td>

[ResourceAuthorization](./data-schema.resourceauthorization.md)


</td><td>

Container for representing the authorization of function resources


</td></tr>
</tbody></table>
