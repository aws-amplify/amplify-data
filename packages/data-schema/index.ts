import * as a from './src';
import { ClientSchema } from './src/ClientSchema';

export { a };

export type { ClientSchema };

export * from './src/runtime/client';
export {
  addSchemaToClient,
  addSchemaToClientWithInstance,
} from './src/runtime';
