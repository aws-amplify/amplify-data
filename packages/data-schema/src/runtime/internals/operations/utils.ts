// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
// import { GraphQLFormattedError } from '@aws-amplify/data-schema-types';

/**
 * Handle errors for list return types (list and index query operations)
 */
export function handleListGraphQlError(error: any) {
  if (error?.errors) {
    // graphql errors pass through
    return {
      ...error,
      data: [],
    } as any;
  } else {
    // non-graphql errors are re-thrown
    throw error;
  }
}

/**
 * Handle errors for singular return types (create, get, update, delete operations)
 */
export function handleSingularGraphQlError(error: any) {
  if (error.errors) {
    // graphql errors pass through
    return {
      ...error,
      data: null,
    } as any;
  } else {
    // non-graphql errors are re-thrown
    throw error;
  }
}

/**
 *
 * @param resolver
 * @returns
 */
export function selfAwareAsync<T>(
  resolver: (promise: Promise<T>) => Promise<T>,
): Promise<T> {
  let resolve: (value: T) => void;
  let reject: (reason?: any) => void;

  const resultPromise = new Promise<T>((res, rej) => {
    resolve = res;
    reject = rej;
  });

  resolver(resultPromise)
    .then((result) => {
      resolve(result);
    })
    .catch((error) => {
      reject(error);
    });

  return resultPromise;
}
