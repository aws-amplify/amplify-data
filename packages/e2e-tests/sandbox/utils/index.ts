import { deploySandbox, deploySandboxWatch } from './deploy';
import { teardownSandbox } from './teardown';
import { updateSchema } from './schemaUtils';

/**
 * Sandbox operations can sometimes be time consuming, so we increase the Jest
 * setup / teardown hook timeouts to 10 minutes.
 */
export const sandboxTimeout: number = 600000;

export const outputsFileName = 'amplify_outputs.json';

export { deploySandbox, deploySandboxWatch };
export { teardownSandbox };
export { updateSchema };
