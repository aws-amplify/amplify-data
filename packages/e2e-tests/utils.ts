import { Amplify } from 'aws-amplify';
import { generateClient } from 'aws-amplify/data';
import type { Schema } from './amplify/data/resource';
import outputs from './amplify_outputs.json';

export const statuses = {
  success: {
    icon: '✅',
    label: 'Success',
  },
  fail: {
    icon: '❌',
    label: 'Fail',
  },
} as const;

const _testCase = (success: boolean): any =>
  success ? statuses.success : statuses.fail;

// TODO: use imported type from `aws-amplify/data` once it's fixed
type Client = ReturnType<typeof configureAmplifyAndGenerateClient>;

export type TestCase = {
  label: string;
  action: (client: Client) => Promise<boolean>;
};

/**
 * Given an array of test cases, runs each test case
 * and asserts that the result is successful
 * @param testCases
 * @param client
 */
export async function runTestCases(testCases: TestCase[]) {
  const client = (global as any).client as Client;

  for (const testCase of testCases) {
    test(`${testCase.label}`, async () => {
      let result: any;

      try {
        const response = await testCase.action(client);
        result = _testCase(response);
      } catch (error) {
        result = _testCase(false);
      }

      console.log(`${result.icon} ${testCase.label}`);
      expect(result.label).toBe(statuses.success.label);
    });
  }
}

/**
 * Configures Amplify and returns API client
 * @returns API client
 */
export function configureAmplifyAndGenerateClient() {
  console.log('configuring Amplify and generating client..');
  Amplify.configure(outputs);

  const client = generateClient<Schema>();

  return client;
}
