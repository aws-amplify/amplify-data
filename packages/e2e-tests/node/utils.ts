import { Amplify } from 'aws-amplify';
import {
  CONNECTION_STATE_CHANGE,
  ConnectionState,
  generateClient,
} from 'aws-amplify/data';
import { Hub, ConsoleLogger } from 'aws-amplify/utils';
import type { Schema } from './amplify/data/resource';
import outputs from './amplify_outputs.json';
// TODO: the dep exists, investigate why it's not being resolved:
import { WebSocket } from 'ws';

// TODO: use imported type from `aws-amplify/data` once it's fixed
export type Client = ReturnType<typeof generateClient<Schema>>;

type ConfigureAmplifyAndGenerateClientParams = {
  disableDebugLogging?: boolean;
  amplifyOptions?: any;
  apiClientOptions?: any;
};

/**
 * Configures Amplify and returns API client
 * @param disableDebugLogging - disables debug logging
 * @param amplifyOptions - options to pass to Amplify.configure
 * @param apiClientOptions - options to pass to generateClient
 * @returns API client
 */
export const configureAmplifyAndGenerateClient = ({
  disableDebugLogging = false,
  amplifyOptions = {},
  apiClientOptions = {},
}: ConfigureAmplifyAndGenerateClientParams): Client => {
  console.log('configuring Amplify and generating client..');
  Amplify.configure(outputs, amplifyOptions);

  /**
   * Debug logs are enabled by default to assist with CI failure debugging.
   * Option to disable is included for local development, in the event that
   * the debug logs are too verbose.
   */
  if (!disableDebugLogging) {
    ConsoleLogger.LOG_LEVEL = 'DEBUG';

    Hub.listen('core', (data: any) => {
      if (data.payload.event === 'configure') {
        console.log('API configuration details:', data.payload.data.API);
      }
    });
  }

  return generateClient<Schema>(apiClientOptions);
};

/**
 * Util that takes model operation response and throws an error if errors
 * are present, or if the data is undefined.
 * @param response - model operation response
 * @param operation - operation name
 * @returns data from response
 */
export async function pause(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

type EstablishWebsocketParams = {
  disableConnectionStateLogging?: boolean;
};

/**
 * Adds the WebSocket API globally.
 * Subscriptions do not work in Node.js environment without the WebSocket API.
 */
export const establishWebsocket = ({
  disableConnectionStateLogging = false,
}: EstablishWebsocketParams): any => {
  (global as any).WebSocket = WebSocket;

  /**
   * Debug logs are enabled by default to assist with CI failure debugging.
   * Option to disable is included for local development, in the event that
   * the debug logs are too verbose.
   */
  if (!disableConnectionStateLogging) {
    // https://docs.amplify.aws/gen1/javascript/build-a-backend/graphqlapi/subscribe-data/#subscription-connection-status-updates
    Hub.listen('api', (data: any) => {
      const { payload } = data;
      if (payload.event === CONNECTION_STATE_CHANGE) {
        const connectionState = payload.data.connectionState as ConnectionState;
        console.log(
          'Connection state has changed. Connection state: ',
          connectionState,
        );
      }
    });
  }
};

export const handleErrorsAndData = <T>(
  response: { data?: T; errors?: any },
  operation: string,
): T => {
  if (response.errors) {
    console.log(`error on ${operation}:`, response.errors);
    throw new Error(JSON.stringify(response.errors));
  }

  if (!response.data) {
    throw new Error(`no response data for ${operation}`);
  }

  return response.data;
};
