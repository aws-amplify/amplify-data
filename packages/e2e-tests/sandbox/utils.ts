import { Amplify } from 'aws-amplify';
import {
  CONNECTION_STATE_CHANGE,
  ConnectionState,
  generateClient,
} from 'aws-amplify/data';
import { Hub, ConsoleLogger } from 'aws-amplify/utils';
import type { Schema } from './amplify/data/resource';
import outputs from './amplify_outputs.json';
import { WebSocket } from 'ws';

// TODO: use imported type from `aws-amplify/data` once it's fixed
export type Client = ReturnType<typeof generateClient<Schema>>;

type ConfigureAmplifyAndGenerateClientParams = {
  enableDebugLogging?: boolean;
  amplifyOptions?: any;
  apiClientOptions?: any;
};

/**
 * Configures Amplify and returns API client
 * @param enableDebugLogging - enables debug logging
 * @param amplifyOptions - options to pass to Amplify.configure
 * @param apiClientOptions - options to pass to generateClient
 * @returns API client
 */
export const configureAmplifyAndGenerateClient = ({
  enableDebugLogging = false,
  amplifyOptions = {},
  apiClientOptions = {},
}: ConfigureAmplifyAndGenerateClientParams): Client => {
  console.log('configuring Amplify and generating client..');
  Amplify.configure(outputs, amplifyOptions);

  if (enableDebugLogging) {
    ConsoleLogger.LOG_LEVEL = 'DEBUG';

    Hub.listen('core', (data: any) => {
      if (data.payload.event === 'configure') {
        console.log('API configuration details:', data.payload.data.API);
      }
    });
  }

  return generateClient<Schema>(apiClientOptions);
};

type AddWebsocketPolyfillParams = {
  enableConnectionStateLogging?: boolean;
};

/**
 * Adds the WebSocket API globally.
 * Subscriptions do not work in Node.js environment without the WebSocket API.
 */
export const addWebsocketPolyfill = ({
  enableConnectionStateLogging = false,
}: AddWebsocketPolyfillParams) => {
  (global as any).WebSocket = WebSocket;

  if (enableConnectionStateLogging) {
    Hub.listen('api', (data: any) => {
      const { payload } = data;
      if (payload.event === CONNECTION_STATE_CHANGE) {
        const connectionState = payload.data.connectionState as ConnectionState;
        console.log(
          'Connection state has changed. Connection state: ',
          connectionState,
        );
      }
    });
  }
};

/**
 * Util that takes model operation response and throws an error if errors
 * are present, or if the data is undefined.
 * @param response - model operation response
 * @param operation - operation name
 * @returns data from response
 */
export const expectDataReturnWithoutErrors = <T>(
  response: { data?: T; errors?: any },
  operation: string,
): T => {
  if (response.errors) {
    console.log(`error on ${operation}:`, response.errors);
    throw new Error(JSON.stringify(response.errors));
  }

  if (!response.data) {
    throw new Error(`no response data for ${operation}`);
  }

  return response.data;
};

/**
 * Function that listens for "Subscription ack" event, and once it receives it,
 * resolves.
 * Jest will timeout if the test does not complete within the default timeout of
 * 5 seconds, so no timeout handling is required.
 */
export const waitForSubscriptionAck = () =>
  new Promise<void>((resolve) => {
    const cancel = Hub.listen('api', async (data: any) => {
      const { payload } = data;
      // TODO: update once we add export:
      if (payload.event === 'Subscription ack') {
        console.log('Subscription ack received:', payload.data);
        // Stop listening for msgs once we receive the ack:
        cancel();
        // Resolve so test execution can continue:
        resolve();
      }
    });
  });
