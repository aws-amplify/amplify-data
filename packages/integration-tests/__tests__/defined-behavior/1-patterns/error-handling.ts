import { a, ClientSchema } from '@aws-amplify/data-schema';
import { Amplify } from 'aws-amplify';
import { buildAmplifyConfig, mockedGenerateClient } from '../../utils';
import { GraphQLError } from 'graphql';

const sampleTodo = {
  __typename: 'Todo',
  id: 'some-id',
  content: 'some content',
  description: 'something something',
  owner: 'some-body',
  done: false,
  updatedAt: '2024-03-01T19:05:44.536Z',
  createdAt: '2024-03-01T18:05:44.536Z',
};

describe('CRUD error handling', () => {
  const authError1 = {
    message: 'Unauthorized',
  } as GraphQLError;

  const authError2 = {
    message: 'Not Authorized to access additionalInfo on type Todo',
  } as GraphQLError;

  /**
   * The following tests demonstrate an error response with an empty `data`
   * result (e.g. `data: {}`).
   * This will occur when, for instance, a customer passes incorrect options
   * when generating the API client.
   * Errors here are caught by the runtime.
   */
  describe('error response with "empty" `data` result', () => {
    // data/resource.ts
    const schema = a.schema({
      Todo: a
        .model({
          content: a.string(),
          description: a.string(),
          done: a.boolean(),
          priority: a.enum(['low', 'medium', 'high']),
        })
        .authorization([a.allow.owner()]),
    });
    type Schema = ClientSchema<typeof schema>;

    afterEach(() => {
      jest.clearAllMocks();
    });

    test('create an item', async () => {
      // #region mocking
      const { spy, innerSpy, generateClient } = mockedGenerateClient([
        {
          data: null,
          errors: [authError1],
        },
      ]);

      // simulated amplifyconfiguration.json
      const config = await buildAmplifyConfig(schema);
      // #endregion mocking

      // App.tsx
      Amplify.configure(config);

      /**
       * Representing an error case where the app is not configured to use IAM,
       * but the customer incorrectly supplies the config anyway. This results
       * in us hitting the following util:
       * https://github.com/aws-amplify/amplify-js/blob/main/packages/api-graphql/src/utils/errors/createGraphQLResultWithError.ts
       */
      const client = generateClient<Schema>({ authMode: 'iam' });

      const { data: newTodo, errors } = await client.models.Todo.create({
        content: 'My new todo',
        done: true,
      });

      // #region assertions
      expect(errors).toEqual(expect.arrayContaining([authError1]));
      expect(newTodo).toBeNull();
      // #endregion assertions
    });

    test('get an item', async () => {
      // #region mocking
      const { spy, innerSpy, generateClient } = mockedGenerateClient([
        {
          data: null,
          errors: [authError1],
        },
      ]);

      // simulated amplifyconfiguration.json
      const config = await buildAmplifyConfig(schema);
      // #endregion mocking

      // App.tsx
      Amplify.configure(config);

      /**
       * Representing an error case where the app is not configured to use IAM,
       * but the customer incorrectly supplies the config anyway. This results
       * in us hitting the following util:
       * https://github.com/aws-amplify/amplify-js/blob/main/packages/api-graphql/src/utils/errors/createGraphQLResultWithError.ts
       */
      const client = generateClient<Schema>({ authMode: 'iam' });

      const { data: todo, errors } = await client.models.Todo.get({
        id: 'some-id',
      });

      // #region assertions
      expect(errors).toEqual(expect.arrayContaining([authError1]));
      expect(todo).toBeNull();
      // #endregion assertions
    });

    test('update an item', async () => {
      // #region mocking
      const { spy, generateClient } = mockedGenerateClient([
        {
          data: null,
          errors: [authError1],
        },
      ]);

      // simulated amplifyconfiguration.json
      const config = await buildAmplifyConfig(schema);
      // #endregion mocking

      // App.tsx
      Amplify.configure(config);

      /**
       * Representing an error case where the app is not configured to use IAM,
       * but the customer incorrectly supplies the config anyway. This results
       * in us hitting the following util:
       * https://github.com/aws-amplify/amplify-js/blob/main/packages/api-graphql/src/utils/errors/createGraphQLResultWithError.ts
       */
      const client = generateClient<Schema>({ authMode: 'iam' });

      const { data: updatedTodo, errors } = await client.models.Todo.update({
        id: 'some_id',
        description: 'Updated description',
      });

      // #region assertions
      expect(errors).toEqual(expect.arrayContaining([authError1]));
      expect(updatedTodo).toBeNull();
      // #endregion assertions
    });

    test('delete an item', async () => {
      // #region mocking
      const { spy, generateClient } = mockedGenerateClient([
        {
          data: null,
          errors: [authError1],
        },
      ]);

      // simulated amplifyconfiguration.json
      const config = await buildAmplifyConfig(schema);
      // #endregion mocking

      // App.tsx
      Amplify.configure(config);

      /**
       * Representing an error case where the app is not configured to use IAM,
       * but the customer incorrectly supplies the config anyway. This results
       * in us hitting the following util:
       * https://github.com/aws-amplify/amplify-js/blob/main/packages/api-graphql/src/utils/errors/createGraphQLResultWithError.ts
       */
      const client = generateClient<Schema>({ authMode: 'iam' });

      const toBeDeletedTodo = {
        id: '123123213',
      };
      const { data: deletedTodo, errors } =
        await client.models.Todo.delete(toBeDeletedTodo);

      // #region assertions
      expect(errors).toEqual(expect.arrayContaining([authError1]));
      expect(deletedTodo).toBeNull();
      // #endregion assertions
    });

    test('list items', async () => {
      // #region mocking
      const { spy, generateClient } = mockedGenerateClient([
        {
          data: [],
          errors: [authError1],
        },
      ]);

      // simulated amplifyconfiguration.json
      const config = await buildAmplifyConfig(schema);
      // #endregion mocking

      // App.tsx
      Amplify.configure(config);

      /**
       * Representing an error case where the app is not configured to use IAM,
       * but the customer incorrectly supplies the config anyway. This results
       * in us hitting the following util:
       * https://github.com/aws-amplify/amplify-js/blob/main/packages/api-graphql/src/utils/errors/createGraphQLResultWithError.ts
       */
      const client = generateClient<Schema>({ authMode: 'iam' });

      const { data: todos, errors } = await client.models.Todo.list({
        id: 'some_id',
      });

      // #region assertions
      expect(errors).toEqual(expect.arrayContaining([authError1]));
      expect(todos).toEqual(expect.arrayContaining([]));
      // #endregion assertions
    });
  });
  /**
   * The following tests demonstrate an error response with a `null` unflattened
   * `data` result, (e.g. `data: createTodo: null`).
   * This can occur when the user attempts to perform an operation that they
   * are not authorized to perform.
   * The error is not caught by the runtime (expected) and `data` is passed
   * through, but the result it not flattened (unexpected).
   */
  describe('error response with `null` result', () => {
    // https://docs.amplify.aws/gen2/build-a-backend/data/mutate-data/

    // data/resource.ts
    const schema = a.schema({
      Todo: a
        .model({
          content: a.string(),
          description: a.string(),
          done: a.boolean(),
          priority: a.enum(['low', 'medium', 'high']),
        })
        .authorization([a.allow.owner()]),
    });
    type Schema = ClientSchema<typeof schema>;

    afterEach(() => {
      jest.clearAllMocks();
    });

    test('create an item', async () => {
      // #region mocking
      const { spy, innerSpy, generateClient } = mockedGenerateClient([
        {
          data: { createTodo: null },
          errors: [authError1],
        },
      ]);

      // simulated amplifyconfiguration.json
      const config = await buildAmplifyConfig(schema);
      // #endregion mocking

      // #region docs code
      // App.tsx
      Amplify.configure(config);
      const client = generateClient<Schema>();

      const { data: newTodo, errors } = await client.models.Todo.create({
        content: 'My new todo',
        done: true,
      });
      // #endregion docs code

      // #region assertions
      expect(errors).toEqual(expect.arrayContaining([authError1]));
      expect(newTodo).toBeNull();
      // #endregion assertions
    });

    test('get an item', async () => {
      // #region mocking
      const { spy, innerSpy, generateClient } = mockedGenerateClient([
        {
          data: { getTodo: null },
          errors: [authError1],
        },
      ]);

      // simulated amplifyconfiguration.json
      const config = await buildAmplifyConfig(schema);
      // #endregion mocking

      // #region docs code
      // App.tsx
      Amplify.configure(config);
      const client = generateClient<Schema>();

      const { data: todo, errors } = await client.models.Todo.get({
        id: 'some-id',
      });
      // #endregion docs code

      // #region assertions
      expect(errors).toEqual(expect.arrayContaining([authError1]));
      expect(todo).toBeNull();
      // #endregion assertions
    });

    test('update an item', async () => {
      // #region mocking
      const { spy, generateClient } = mockedGenerateClient([
        {
          data: { updateTodo: null },
          errors: [authError1],
        },
      ]);

      // simulated amplifyconfiguration.json
      const config = await buildAmplifyConfig(schema);
      // #endregion mocking

      // #region docs code
      // App.tsx
      Amplify.configure(config);
      const client = generateClient<Schema>();
      const { data: updatedTodo, errors } = await client.models.Todo.update({
        id: 'some_id',
        description: 'Updated description',
      });
      // #endregion docs code

      // #region assertions
      expect(errors).toEqual(expect.arrayContaining([authError1]));
      expect(updatedTodo).toBeNull();
      // #endregion assertions
    });

    test('delete an item', async () => {
      // #region mocking
      const { spy, generateClient } = mockedGenerateClient([
        {
          data: { deleteTodo: null },
          errors: [authError1],
        },
      ]);

      // simulated amplifyconfiguration.json
      const config = await buildAmplifyConfig(schema);
      // #endregion mocking

      // #region docs code
      // App.tsx
      Amplify.configure(config);
      const client = generateClient<Schema>();
      const toBeDeletedTodo = {
        id: '123123213',
      };
      const { data: deletedTodo, errors } =
        await client.models.Todo.delete(toBeDeletedTodo);
      // #endregion docs code

      // #region assertions
      expect(errors).toEqual(expect.arrayContaining([authError1]));
      expect(deletedTodo).toBeNull();
      // #endregion assertions
    });

    test('list items', async () => {
      // #region mocking
      const { spy, generateClient } = mockedGenerateClient([
        {
          data: { listTodo: null },
          errors: [authError1],
        },
      ]);

      // simulated amplifyconfiguration.json
      const config = await buildAmplifyConfig(schema);
      // #endregion mocking

      // #region docs code
      // App.tsx
      Amplify.configure(config);
      const client = generateClient<Schema>();

      const { data: todos, errors } = await client.models.Todo.list({
        id: 'some_id',
      });
      // #endregion docs code

      // #region assertions
      expect(errors).toEqual(expect.arrayContaining([authError1]));
      expect(todos).toEqual(expect.arrayContaining([]));
      // #endregion assertions
    });
  });
  /**
   * The following tests demonstrate an error response with a complete and
   * unflattened `data` result (e.g. `data: createTodo: {id: '123', ...etc }`).
   * This can occur when, for instance, a selection set is provided that
   * includes fields that the user is not authorized to access.
   * The error is not caught by the runtime (expected) and `data` is passed
   * through, but the result it not flattened (unexpected).
   */
  describe('error response with populated `data` result', () => {
    // data/resource.ts

    // Schema that allows public users create-only access to `Note`:
    const schema = a.schema({
      Todo: a
        .model({
          content: a.string(),
          description: a.string(),
          additionalInfo: a.hasOne('Note'),
        })
        .authorization([a.allow.public()]),
      Note: a
        .model({
          content: a.string(),
        })
        .authorization([a.allow.public().to(['create']), a.allow.owner()]),
    });

    type Schema = ClientSchema<typeof schema>;

    afterEach(() => {
      jest.clearAllMocks();
    });

    test.only('get an item', async () => {
      // #region mocking
      const { spy, generateClient } = mockedGenerateClient([
        {
          data: { getTodo: sampleTodo },
          errors: [authError2],
        },
      ]);

      // simulated amplifyconfiguration.json
      const config = await buildAmplifyConfig(schema);
      // #endregion mocking

      // App.tsx
      Amplify.configure(config);
      const client = generateClient<Schema>();
      const { data: getTodo, errors } = await client.models.Todo.get(
        {
          id: 'some_id',
        },
        {
          selectionSet: ['id', 'content', 'additionalInfo.*'],
        },
      );

      // #region assertions
      expect(errors).toEqual(expect.arrayContaining([authError2]));
      expect(getTodo).toEqual({ getTodo: sampleTodo });
      // #endregion assertions
    });
    test('list an item', async () => {
      // #region mocking
      const { spy, generateClient } = mockedGenerateClient([
        {
          data: { listTodos: [sampleTodo] },
          errors: [authError2],
        },
      ]);

      // simulated amplifyconfiguration.json
      const config = await buildAmplifyConfig(schema);
      // #endregion mocking

      // App.tsx
      Amplify.configure(config);
      const client = generateClient<Schema>();
      const { data, errors } = await client.models.Todo.list({
        selectionSet: ['id', 'content', 'additionalInfo.*'],
      });

      // #region assertions
      expect(errors).toEqual(expect.arrayContaining([authError2]));
      expect(data).toEqual({ listTodos: [sampleTodo] });
      // #endregion assertions
    });
  });
});
