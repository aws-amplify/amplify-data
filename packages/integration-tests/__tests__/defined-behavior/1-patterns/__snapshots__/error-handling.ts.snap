// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CRUD error handling error response with "empty" \`data\` result create an item 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "mutation($input: CreateTodoInput!) { createTodo(input: $input) { id content description done priority createdAt updatedAt owner } }",
      "variables": {
        "input": {
          "content": "My new todo",
          "done": true,
        },
      },
    },
    {},
  ],
]
`;

exports[`CRUD error handling error response with "empty" \`data\` result delete an item 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "mutation($input: DeleteTodoInput!) { deleteTodo(input: $input) { id content description done priority createdAt updatedAt owner } }",
      "variables": {
        "input": {
          "id": "123123213",
        },
      },
    },
    {},
  ],
]
`;

exports[`CRUD error handling error response with "empty" \`data\` result update an item 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "mutation($input: UpdateTodoInput!) { updateTodo(input: $input) { id content description done priority createdAt updatedAt owner } }",
      "variables": {
        "input": {
          "description": "Updated description",
          "id": "some_id",
        },
      },
    },
    {},
  ],
]
`;

exports[`CRUD error handling error response with populated \`data\` result get an item 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "query($id: ID!) { getTodo(id: $id) { id content additionalInfo { id content createdAt updatedAt owner } } }",
      "variables": {
        "id": "some_id",
      },
    },
    {},
  ],
]
`;

exports[`CRUD error handling error response with populated \`data\` result list an item 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "query($filter: ModelTodoFilterInput,$limit: Int,$nextToken: String) { listTodos(filter: $filter,limit: $limit,nextToken: $nextToken) { items { id content additionalInfo { id content createdAt updatedAt owner } } nextToken __typename } }",
      "variables": {},
    },
    {},
  ],
]
`;
